### PR TITLE
A new example to reduce mock duplicate creation in AsyncQueryForwardingServletTest.java

### PR DESCRIPTION
**Pull request by gzhao9**
#### Description
Fixes #xxxx

#### Background (content of the issue)

---

Hello Druid Team,

Firstly, I want to extend my gratitude for your insights shared in the review comments of `#15980` and `#15187`. 

We respect the concerns about 'reducing only one line in each test' and its potential impact on readability, in light of this, I have prepared a new example that I believe aligns better with both the project's standards and the constructive guidance you've provided.

This new example **reduces 5-6 lines per test case**, and the total reduction is 33 lines in its file. I believe that this level of code reduction does not compromise test readability or integrity. Could it present a more compelling case for making these changes?

 For instance, in test class [AsyncQueryForwardingServletTest](https://github.com/apache/druid/blob/daf03939a92763a104dc593de2057b7d9c28d7d8/services/src/test/java/org/apache/druid/server/AsyncQueryForwardingServletTest.java#L404C1-L409C71), the following mock setup:

```java
errorMessage = "test exception message"; 
mockMapper = Mockito.mock(ObjectMapper.class);
request = Mockito.mock(HttpServletRequest.class);
response = Mockito.mock(HttpServletResponse.class);
outputStream = Mockito.mock(ServletOutputStream.class);
Mockito.when(response.getOutputStream()).thenReturn(outputStream);
```

is repeated in 6 test cases. To reduce this repetition and streamline the test setup process, I proposed moving these mock object creations to a `@Before` method. Like following:
```
java
 @Before
  public void setup() throws Exception
    lifecycle.start();
    ClientHolder holder = injector.getInstance(ClientHolder.class);
    client = holder.getClient();
    errorMessage = "test exception message"; 
    mockMapper = Mockito.mock(ObjectMapper.class);
    request = Mockito.mock(HttpServletRequest.class);
    response = Mockito.mock(HttpServletResponse.class);
    outputStream = Mockito.mock(ServletOutputStream.class);
    Mockito.when(response.getOutputStream()).thenReturn(outputStream);
  }
```
You can see the detailed modifications [here.](https://github.com/gzhao9/druid/pull/2/files)
With this modification, **each test case is reduced by 5-6 lines.** A total of 33 lines have been reduced. And does not reduce readability. could Druid consider this modification?

---

This PR addresses the issue #xxxx 

#### Description
This PR is aimed at resolving issue #xxxx, which involves duplicate creation when creating objects in the `AsyncQueryForwardingServletTest` class. By suggestions from `#15980` and `#15187`, This issue achieves a reduction **of 5-6 lines per test case and a total of 33 lines across the file.**

#### Solution
To reduce this repetition and streamline the test setup process, I proposed moving the duplicate creation mock object to a `@Before` method. 

#### Key changed in this PR
- **Added**: `@Before` method in `AsyncQueryForwardingServletTest` for centralized mock object creation.
- **Refactored**: Reduced redundancy in mock object creation across 6 test cases in `AsyncQueryForwardingServletTest`, achieving a **reduction of 5-6 lines per test case** and a **total of 33 lines** across the file.

Thank you for considering this contribution. I look forward to your feedback and further discussions.

Best regards,

### CODE REVIEW

File: services/src/test/java/org/apache/druid/server/AsyncQueryForwardingServletTest.java [hunk 1]
```diff
@@ -127,7 +127,11 @@ public class AsyncQueryForwardingServletTest extends BaseJettyTest
 {
   private static int port1;
   private static int port2;
-
+  private String errorMessage;
+  private ObjectMapper mockMapper;
+  private HttpServletRequest request;
+  private HttpServletResponse response;
+  private ServletOutputStream outputStream;
```
Comment (kfaraz) [inline services/src/test/java/org/apache/druid/server/AsyncQueryForwardingServletTest.java:134]: Please add a line after this.

```suggestion
  private ServletOutputStream outputStream;
  
```

File: services/src/test/java/org/apache/druid/server/AsyncQueryForwardingServletTest.java [hunk 2]
```diff
@@ -127,7 +127,11 @@ public class AsyncQueryForwardingServletTest extends BaseJettyTest
 {
   private static int port1;
   private static int port2;
-
+  private String errorMessage;
```
Comment (kfaraz) [inline services/src/test/java/org/apache/druid/server/AsyncQueryForwardingServletTest.java:130]: The line above was also good to have a visual separation of the static and non-static fields.
```suggestion

  private String errorMessage;
```

**Review by kfaraz — CHANGES_REQUESTED** [review]
@gzhao9 , thanks for your interest in Druid tests.
Looking at these tests, there is in fact a lot more that you can common out.
I don't think it makes a lot of sense to initialize these fields in the `setup` method as there are several other tests which don't use these fields.
Instead you should create a private method which does all of this work and is called from all of these tests.

It may look something like this:

```java
private <T extends Throwable> void assertExceptionHandled(
      ServerConfig serverConfig,
      T testException,
      String expectedExceptionMessage,
      String expectedExceptionClass,
      ExceptionHandler<T> handler
  ) throws Exception
  {
    ObjectMapper mockMapper = Mockito.mock(ObjectMapper.class);
    HttpServletRequest request = Mockito.mock(HttpServletRequest.class);
    HttpServletResponse response = Mockito.mock(HttpServletResponse.class);
    ServletOutputStream outputStream = Mockito.mock(ServletOutputStream.class);
    Mockito.when(response.getOutputStream()).thenReturn(outputStream);
    final AsyncQueryForwardingServlet servlet = new AsyncQueryForwardingServlet(
        new MapQueryToolChestWarehouse(ImmutableMap.of()),
        mockMapper,
        TestHelper.makeSmileMapper(),
        null,
        null,
        null,
        new NoopServiceEmitter(),
        new NoopRequestLogger(),
        new DefaultGenericQueryMetricsFactory(),
        new AuthenticatorMapper(ImmutableMap.of()),
        new Properties(),
        serverConfig
    );
    Mockito.when(request.getAttribute(AuthConfig.DRUID_AUTHENTICATION_RESULT))
           .thenReturn(new AuthenticationResult("userA", "basic", "basic", null));

    handler.handle(servlet, request, response, mockMapper, testException);

    ArgumentCaptor<Exception> captor = ArgumentCaptor.forClass(Exception.class);
    Mockito.verify(mockMapper).writeValue(ArgumentMatchers.eq(outputStream), captor.capture());
    Assert.assertTrue(captor.getValue() instanceof QueryException);
    Assert.assertEquals("Unknown exception", ((QueryException) captor.getValue()).getErrorCode());
    Assert.assertEquals(expectedExceptionMessage, captor.getValue().getMessage());
    Assert.assertEquals(expectedExceptionClass, ((QueryException) captor.getValue()).getErrorClass());
    Assert.assertNull(((QueryException) captor.getValue()).getHost());
  }

  private interface ExceptionHandler<T extends Throwable>
  {
    void handle(
        AsyncQueryForwardingServlet servlet,
        HttpServletRequest request,
        HttpServletResponse response,
        ObjectMapper mapper,
        T exception
    ) throws Exception;
  }
```

and the tests would be modified to something like

```java
@Test
  public void testHandleQueryParseExceptionWithFilterEnabled() throws Exception
  {
    final ServerConfig serverConfig = new ServerConfig()
    {
      @Override
      public boolean isShowDetailedJettyErrors()
      {
        return true;
      }

      @Override
      public ErrorResponseTransformStrategy getErrorResponseTransformStrategy()
      {
        return new AllowedRegexErrorResponseTransformStrategy(ImmutableList.of());
      }
    };
    assertExceptionHandled(
        serverConfig,
        new IOException("test exception message"),
        (servlet, request, response, mapper, exception)
            -> servlet.handleQueryParseException(request, response, mapper, exception, false)
    );
}    
```

This would make the tests much more concise and readable.
Please let me know if it makes sense.

**Comment by gzhao9** [comment]
@kfaraz Thank you very much for your suggestions, your suggested teams are very useful.
I have made the changes to the 6test as you suggested.
The main modification was to create a private method and put the duplicate parts of the 6 tests in this private method.

I used a similar approach to what you suggested.
But the difference is that, considering using it in 6 different test cases, so instead of having this method execute assert statements, I let this method return `ArgumentCaptor<Exception>`
Like the following.
```java
 private ArgumentCaptor<Exception> getArgumentCaptor(ServerConfig serverConfig,String errorMessage,boolean haverequest) throws IOException {
    ObjectMapper mockMapper = Mockito.mock(ObjectMapper.class);
    HttpServletResponse response = Mockito.mock(HttpServletResponse.class);
    ServletOutputStream outputStream = Mockito.mock(ServletOutputStream.class);
    Mockito.when(response.getOutputStream()).thenReturn(outputStream);
    final AsyncQueryForwardingServlet servlet = new AsyncQueryForwardingServlet(
            new MapQueryToolChestWarehouse(ImmutableMap.of()),
            mockMapper,
            TestHelper.makeSmileMapper(),
            null,
            null,
            null,
            new NoopServiceEmitter(),
            new NoopRequestLogger(),
            new DefaultGenericQueryMetricsFactory(),
            new AuthenticatorMapper(ImmutableMap.of()),
            new Properties(),
            serverConfig
    );
    Exception testException = new IllegalStateException(errorMessage);

    if(haverequest){
      HttpServletRequest request = Mockito.mock(HttpServletRequest.class);
      Mockito.when(request.getAttribute(AuthConfig.DRUID_AUTHENTICATION_RESULT)).thenReturn(new AuthenticationResult("userA", "basic", "basic", null));
      IOException testException2 = new IOException(errorMessage);
      servlet.handleQueryParseException(request, response, mockMapper, testException2, false);
    }
    else {
      servlet.handleException(response, mockMapper, testException);
    }

    ArgumentCaptor<Exception> captor = ArgumentCaptor.forClass(Exception.class);
    Mockito.verify(mockMapper).writeValue(ArgumentMatchers.eq(outputStream), captor.capture());

    return captor;
  }
```
This way we only have to focus on whether or not the request was created.

Before calling this method, the test case looks like this.
```java
 @Test
  public void testHandleQueryParseExceptionWithFilterDisabled() throws Exception
  {
    String errorMessage = "test exception message";
    ObjectMapper mockMapper = Mockito.mock(ObjectMapper.class);
    HttpServletRequest request = Mockito.mock(HttpServletRequest.class);
    HttpServletResponse response = Mockito.mock(HttpServletResponse.class);
    ServletOutputStream outputStream = Mockito.mock(ServletOutputStream.class);
    Mockito.when(response.getOutputStream()).thenReturn(outputStream);
    final AsyncQueryForwardingServlet servlet = new AsyncQueryForwardingServlet(
        new MapQueryToolChestWarehouse(ImmutableMap.of()),
        mockMapper,
        TestHelper.makeSmileMapper(),
        null,
        null,
        null,
        new NoopServiceEmitter(),
        new NoopRequestLogger(),
        new DefaultGenericQueryMetricsFactory(),
        new AuthenticatorMapper(ImmutableMap.of()),
        new Properties(),
        new ServerConfig()
    );
    Mockito.when(request.getAttribute(AuthConfig.DRUID_AUTHENTICATION_RESULT)).thenReturn(new AuthenticationResult("userA", "basic", "basic", null));
    IOException testException = new IOException(errorMessage);
    servlet.handleQueryParseException(request, response, mockMapper, testException, false);
    ArgumentCaptor<Exception> captor = ArgumentCaptor.forClass(Exception.class);
    Mockito.verify(mockMapper).writeValue(ArgumentMatchers.eq(outputStream), captor.capture());
    Assert.assertTrue(captor.getValue() instanceof QueryException);
    Assert.assertEquals("Unknown exception", ((QueryException) captor.getValue()).getErrorCode());
    Assert.assertEquals(errorMessage, captor.getValue().getMessage());
    Assert.assertEquals(IOException.class.getName(), ((QueryException) captor.getValue()).getErrorClass());
  }
```

With modifications, the test looks like this.
```java
  @Test
  public void testHandleQueryParseExceptionWithFilterDisabled() throws Exception
  {
    String errorMessage = "test exception message";
    boolean haveRequest=true;
    ServerConfig serverConfig= new ServerConfig();
    ArgumentCaptor<Exception> captor=getArgumentCaptor(serverConfig,errorMessage,haveRequest);
    Assert.assertTrue(captor.getValue() instanceof QueryException);
    Assert.assertEquals("Unknown exception", ((QueryException) captor.getValue()).getErrorCode());
    Assert.assertEquals(errorMessage, captor.getValue().getMessage());
    Assert.assertEquals(IOException.class.getName(), ((QueryException) captor.getValue()).getErrorClass());
  }
```

This change appears in six test cases, which can make each test case more concise and easy to read.
You can get the changes [here](https://github.com/gzhao9/druid/pull/2/files)
Again, thank you very much for your suggestions. I would be very grateful if you let me know if this change is worthwhile for submission PR.

**Comment by kfaraz** [comment]
This looks good, @gzhao9 ! Could you please create a PR against apache/druid? It would be easier to review there.

**Comment by kfaraz** [comment]
This looks good, @gzhao9 ! Could you please create a PR against Apache Druid? It would be easier to review there.

**Comment by gzhao9** [comment]
> This looks good, @gzhao9 ! Could you please create a PR against apache/druid? It would be easier to review there.

Ok, thanks a lot, I will create a PR!

**Pull request closed without merging**